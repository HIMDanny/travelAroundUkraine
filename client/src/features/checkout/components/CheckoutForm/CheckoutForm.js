import React, { useEffect,memo } from 'react';
// Formik
import { Formik, Form } from 'formik';
import { LocalizationProvider } from '@mui/x-date-pickers';
import { AdapterDateFns } from '@mui/x-date-pickers/AdapterDateFns';
import { enAU } from 'date-fns/locale';
// MUI Components
import { Stepper, Step, StepLabel } from '@mui/material';
// Redux
import { useSelector, useDispatch, shallowEqual } from 'react-redux';
import { fetchUserInfo } from '../../../../store/slices/userSlice/userSlice';
import {
  increaseStep,
  decreaseStep,
  createNewOrder,
  setIsModalOpen,
} from '../../../../store/slices/orderSlice/orderSlice';
// Child Forms and model
import { UserDetailsForm, ShippingAddressForm, PaymentForm, PaymentSuccess, CheckoutSummary } from '..';
import FormBttnContainer from '../../../../components/FormBttnContainer/FormBttnContainer';
import { validationSchema } from '../../data';
import setInitialValue from '../../utils/setInitialValue';

const steps = ['User Details', 'Shipping Address', 'Payment Details'];

const CheckoutForm = () => {
  const { currentStep, isLoading } = useSelector((state) => state.order, shallowEqual);
  const isLogin = useSelector((store) => store.user.isLogin);
  const userData = useSelector((store) => store.user.userData);

  const dispatch = useDispatch();

  const lastStep = steps.length - 1;
  const currentValidationSchema = validationSchema[currentStep];

  const goToNextStep = () => {
    dispatch(increaseStep());
  };

  const goToPrevStep = () => {
    dispatch(decreaseStep());
  };

  const formSubmitHandler = async (values) => {
    if (currentStep !== lastStep) {
      goToNextStep();
      return;
    }

    try {
      await dispatch(createNewOrder(values)).unwrap();

      dispatch(setIsModalOpen(true));
    } catch (e) {
      console.error(e);
    }
  };

  useEffect(() => {
    if (isLogin) {
      dispatch(fetchUserInfo());
    }
  }, []);

  return (
    <>
      <Stepper activeStep={currentStep} alternativeLabel sx={{ mb: '100px' }}>
        {steps.map((label) => (
          <Step key={label}>
            <StepLabel>{label}</StepLabel>
          </Step>
        ))}
      </Stepper>

      {currentStep === steps.length ? (
        <CheckoutSummary />
      ) : (
        <Formik
          initialValues={setInitialValue(isLogin, userData)}
          validationSchema={currentValidationSchema}
          onSubmit={formSubmitHandler}
        >
          {({ isSubmitting }) => (
            <LocalizationProvider dateAdapter={AdapterDateFns} adapterLocale={enAU}>
              <Form>
                {currentStep === 0 && <UserDetailsForm />}

                {currentStep === 1 && <ShippingAddressForm />}

                {currentStep === 2 && <PaymentForm />}

                <FormBttnContainer
                  currentStep={currentStep}
                  lastStep={lastStep}
                  isLoading={isLoading}
                  isSubmitting={isSubmitting}
                  goToPrevStep={goToPrevStep}
                  bttnText="Continue"
                  bttnLastStepText="Confirm"
                />
              </Form>
            </LocalizationProvider>
          )}
        </Formik>
      )}

      <PaymentSuccess />
    </>
  );
};

export default memo(CheckoutForm);
