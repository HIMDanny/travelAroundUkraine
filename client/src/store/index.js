import { configureStore } from '@reduxjs/toolkit';
import userReducer from './slices/userSlice/userSlice';
import tourReducer from './slices/tourSlice/tourSlice';
import catalogueReducer from './slices/catalogueSlice/catalogueSlice';
import cartReducer from './slices/cartSlice/cartSlice';
import filterReducer from './slices/filterSlice/filterSlice';
import orderSlice from './slices/orderSlice/orderSlice';
import searchBarReducer from './slices/searchBarSlice/searchBarSlice';

const store = configureStore({
  reducer: {
    user: userReducer,
    tour: tourReducer,
    catalogue: catalogueReducer,
    cart: cartReducer,
    filter: filterReducer,
    order: orderSlice,
    searchBar: searchBarReducer,
  },
});

export default store;
