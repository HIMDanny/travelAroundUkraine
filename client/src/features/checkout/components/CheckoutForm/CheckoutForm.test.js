/* eslint-disable react/prop-types */
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import store from '../../../../store';
import CheckoutForm from './CheckoutForm';

jest.mock('@mui/material/Stepper', () => ({ children }) => <div>{children}</div>);
jest.mock('@mui/material/Step', () => ({ children }) => <ul>{children}</ul>);
jest.mock('@mui/material/StepLabel', () => ({ children }) => <li>{children}</li>);
jest.mock('@mui/material/Button', () => ({ children }) => <button>{children}</button>);
jest.mock('../../components/CheckoutSummary/CheckoutSummary', () => () => <div>CHECKOUT SUMMARY</div>);
jest.mock('../../components/PaymentSuccess/PaymentSuccess', () => () => <div>PAYMENT SUCCESS</div>);
jest.mock('../../components/UserDetailsForm/UserDetailsForm', () => () => <div>USER DETAILS FORM</div>);

describe('CheckoutForm snapshot test', () => {
  test('should CheckoutForm match snapshot', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <CheckoutForm />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
