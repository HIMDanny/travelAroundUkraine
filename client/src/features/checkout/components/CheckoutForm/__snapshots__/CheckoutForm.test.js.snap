// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckoutForm snapshot test should CheckoutForm match snapshot 1`] = `
<DocumentFragment>
  <div>
    <ul>
      <li>
        User Details
      </li>
    </ul>
    <ul>
      <li>
        Shipping Address
      </li>
    </ul>
    <ul>
      <li>
        Payment Details
      </li>
    </ul>
  </div>
  <form
    action="#"
  >
    <div>
      USER DETAILS FORM
    </div>
    <div
      class="MuiBox-root css-1s9g7sl"
    >
      <button>
        Continue
      </button>
    </div>
  </form>
  <div>
    PAYMENT SUCCESS
  </div>
</DocumentFragment>
`;
